### PR TITLE
Release - allow deletion of a webauthn credential or legacy u2f

### PR DESCRIPTION
**Pull request by Baggerone**

### CODE REVIEW

File: server/main.go [hunk 1]
```diff
@@ -75,6 +75,16 @@ var routes = []route{
...
 		mfa.DeleteUser,
 	},
+	{ // This expects a path param that is the id that was previously returned
+		// as the key_handle_hash from the FinishRegistration call.
+		// Alternatively, if the id param indicates that a legacy U2F key should be removed
+		//	 (e.g. by matching the string "u2f")
+		//   then that user is saved with all of its legacy u2f fields blanked out.
+		"DeleteCredential",
+		"DELETE",
+		fmt.Sprintf("/webauthn/credential/{%s}", mfa.IDParam),
```
Comment (jason-jackson) [inline server/main.go:85]: Probably delete this

File: user.go [hunk 2]
```diff
@@ -117,6 +130,59 @@ func (u *DynamoUser) saveNewCredential(credential webauthn.Credential) error {
...
+	return u.encryptAndStoreCredentials()
+}
+
+// DeleteCredential expects a hashed-encoded credential id.
+//  It finds a matching credential for that user and saves the user
+//    without that credential included.
+//  Alternatively, if the given credential id indicates that a legacy U2F key should be removed
+//	 (e.g. by matching the string "u2f")
+//    then that user is saved with all of its legacy u2f fields blanked out.
+func (u *DynamoUser) DeleteCredential(credIDHash string) (error, int) {
```
Comment (jason-jackson) [inline user.go:142]: I'm more used to error being last if multiple return values.

File: suit_test.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package mfa
+
```
Comment (forevermatt) [inline suit_test.go:2]: Did you mean to name this file `suite_test.go` rather than `suit_test.go`?

**Review by forevermatt — APPROVED** [review]
I'm not sure how much my approval of Go code is worth (😉), especially of `webauthn_test.go`, which I ended up mostly just skimming.

File: server/main.go (hunk 1 above)
Comment (jason-jackson) [reply server/main.go:85]: @forevermatt just the Sprintf, I think it was just used for testing.

**Review by jason-jackson — APPROVED** [review]

File: server/main.go (hunk 1 above)
Comment (Baggerone) [reply server/main.go:85]: That's the `pattern` value for that route.

File: webauthn.go [hunk 4]
```diff
@@ -141,6 +144,31 @@ func DeleteUser(w http.ResponseWriter, r *http.Request) {
 	jsonResponse(w, nil, http.StatusNoContent)
 }
 
+func DeleteCredential(w http.ResponseWriter, r *http.Request) {
+	user, err := getUserFromContext(r)
+	if err != nil {
+		jsonResponse(w, err, http.StatusBadRequest)
```
Comment (Schparky) [inline webauthn.go:150]: Should this be `http.StatusInternalServerError`?

**Review by Schparky — APPROVED** [review]

**Comment by Schparky** [comment]
Should the PR title have the word "Release" in in?

File: fixtures_test.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/duo-labs/webauthn/webauthn"
+)
+
+type fixtureConfig struct {
+	Storage       *Storage
+	WebAuthClient *webauthn.WebAuthn
```
[the remark this answers is not in the document]
Comment (Baggerone) [reply fixtures_test.go:11]: Looks like that struct is never used anyway. I'll get rid of it.

File: testutils.go [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/duo-labs/webauthn/protocol"
+	"github.com/duo-labs/webauthn/protocol/webauthncbor"
+	"github.com/duo-labs/webauthn/protocol/webauthncose"
+)
+
+const (
+	localAppID = "http://localhost"
+
+	// 65 = AT(64) + UP(1)
```
[the remark this answers is not in the document]
Comment (Baggerone) [reply testutils.go:28]: I'll add more comments to clarify it.

File: webauthn.go (hunk 4 above)
Comment (Baggerone) [reply webauthn.go:150]: I don't think so.  In any case, that matches all the other places getUserFromContext is called.

**Pull request merged**